stable_baselines/common/callbacks.py

```diff
@@ -0,0 +1,371 @@
+import os
+from abc import ABC
+import warnings
+import typing
+from typing import Union, List, Dict, Any, Optional
+
+import gym
+import numpy as np
+
+from stable_baselines.common.vec_env import VecEnv, sync_envs_normalization, DummyVecEnv
+from stable_baselines.common.evaluation import evaluate_policy
+from stable_baselines import logger
+
+if typing.TYPE_CHECKING:
+    from stable_baselines.common.base_class import BaseRLModel  # pytype: disable=pyi-error
+
+
+class BaseCallback(ABC):
+    """
+    Base class for callback.
+
+    :param verbose: (int)
+    """
+    def __init__(self, verbose: int = 0):
+        super(BaseCallback, self).__init__()
+        # The RL model
+        self.model = None  # type: Optional[BaseRLModel]
+        # An alias for self.model.get_env(), the environment used for training
+        self.training_env = None  # type: Union[gym.Env, VecEnv, None]
+        # Number of time the callback was called
+        self.n_calls = 0  # type: int
+        # n_envs * n times env.step() was called
+        self.num_timesteps = 0  # type: int
+        self.verbose = verbose
+        self.locals = None  # type: Optional[Dict[str, Any]]
+        self.globals = None  # type: Optional[Dict[str, Any]]
+        self.logger = None  # type: Optional[logger.Logger]
+        # Sometimes, for event callback, it is useful
+        # to have access to the parent object
+        self.parent = None  # type: Optional[BaseCallback]
+
+    # Type hint as string to avoid circular import
+    def init_callback(self, model: 'BaseRLModel') -> None:
+        """
+        Initialize the callback by saving references to the
+        RL model and the training environment for convenience.
+        """
+        self.model = model
+        self.training_env = model.get_env()
+        self.logger = logger.Logger.CURRENT
+        self._init_callback()
+
+    def _init_callback(self) -> None:
+        pass
+
+    def on_training_start(self, locals_: Dict[str, Any], globals_: Dict[str, Any]) -> None:
+        # Those are reference and will be updated automatically
+        self.locals = locals_
+        self.globals = globals_
+        self._on_training_start()
+
+    def _on_training_start(self) -> None:
+        pass
+
+    def on_rollout_start(self) -> None:
+        self._on_rollout_start()
+
+    def _on_rollout_start(self) -> None:
+        pass
+
+    def _on_step(self) -> bool:
+        """
+        :return: (bool) If the callback returns False, training is aborted early.
+        """
+        return True
+
+    def on_step(self) -> bool:
+        """
+        This method will be called by the model after each call to `env.step()`.
+        This is the equivalent to the callback function in previous versions.
```

Comment:
Did this not change? Previously callbacks were called at e.g. after training steps, but now it is really after `env.step()`?

Reply:
Previously, it was a mess, nothing was consistent...

Reply:
Fair point. I would drop this comment, though, just so people do not get mixed ideas of what `on_step` actually does :)